packages/lisk-p2p/src/peer_selector.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright © 2018 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+import { NotEnoughPeersError } from './errors';
+import { Peer } from './peer';
+
+// For future extension of return type
+export interface PeerReturnType {
+	readonly peers: ReadonlyArray<Peer>;
+}
+export interface PeerOptions {
+	readonly [key: string]: string | number;
+}
+/* tslint:disable: readonly-keyword*/
+interface Histogram {
+	[key: number]: number;
+}
+interface HistogramValues {
+	height: number;
+	histogram: Histogram;
+	max: number;
+}
+/* tslint:enable: readonly-keyword */
+export const selectPeers = (
+	peers: ReadonlyArray<Peer>,
+	selectionParams: PeerOptions = { lastBlockHeight: 0 },
+	numOfPeers?: number,
+): PeerReturnType => {
+	const filteredPeers = peers.filter(
+		// Remove unreachable peers or heights below last block height
+		(peer: Peer) => peer.height >= selectionParams.lastBlockHeight,
+	);
+
+	if (filteredPeers.length === 0) {
+		return { peers: [] };
+	}
+
+	// Order peers by descending height
+	const sortedPeers = filteredPeers.sort((a, b) => b.height - a.height);
+
+	const aggregation = 2;
+
+	const calculatedHistogramValues = sortedPeers.reduce(
+		(
+			histogramValues: HistogramValues = {
+				height: 0,
+				histogram: {},
+				max: -1,
+			},
+			peer: Peer,
+		) => {
+			const val = Math.floor(peer.height / aggregation) * aggregation;
+			histogramValues.histogram[val] =
+				(histogramValues.histogram[val] ? histogramValues.histogram[val] : 0) +
+				1;
+			if (histogramValues.histogram[val] > histogramValues.max) {
+				histogramValues.max = histogramValues.histogram[val];
+				histogramValues.height = val;
+			}
+
+			return histogramValues;
+		},
+		{ height: 0, histogram: {}, max: -1 },
+	);
+
+	// Perform histogram cut of peers too far from histogram maximum
+	const processedPeers = sortedPeers.filter(
+		peer =>
+			peer &&
+			Math.abs(
+				(calculatedHistogramValues ? calculatedHistogramValues.height : 0) -
+					peer.height,
+			) <
+				aggregation + 1,
+	);
```

Comment:
Spacing and indentation looks wrong here.

Reply:
It was done by prettify, but let me check if I can avoid prettify for this expression

Reply:
😢